Avoid using global variables for configuration

alebedev87 opened the pull request:
**Current situation**
The setup of the webhook checks for OpenShift API to set a global `api/v1alpha1` package variable which indicates whether it's openshift or not. In the same place the infrastructure config is retrieved and and another global variable for `PlatformStatus` is set in the same package. The rest of the code uses this to switchcase between pure Kubernetes and OpenShift installations.

_Problem 1_: platform detection and status have no logical adherence to the API wehbook. This is something which has to be done once but the webhook is not optimal place for this. 
_Problem 2_: testing with global variables is more difficult as the functions depend on external variants which are not easy to set and even more difficult to be reset to the previous values.

**Proposal**
Move the platform specific things into the operator's config: `pkg/operator/config#Config`. The platform gets detected in main (only Kube client is needed), the platforms status gets retrieved at the operator initialization (we already know the platform and can use the specific client). Then these data is propagated to the places where it's needed.

_Result_: the things which have to be done once are not bound to the webhook setup, fewer dependencies on the global variables. Slightly (a couple of %) better code coverage.
 
 
 **Bonus**
 Discovered by e2e testing of the proposal: credentials secret specified in `ExternalDNS` CR is not used when the platform is OpenShift. Fix this to give the priority to explicitly specified secret.

Miciah commented on pkg/operator/controller/externaldns/pod.go:
```diff
@@ -345,8 +335,9 @@ func (b *externalDNSContainerBuilder) fillGCPFields(container *corev1.Container)
 		}
 	} else {
-		container.Args = append(container.Args, fmt.Sprintf("--google-project=%s", operatorv1alpha1.PlatformStatus.GCP.ProjectID))
+		container.Args = append(container.Args, fmt.Sprintf("--google-project=%s", b.platformStatus.GCP.ProjectID))
```
This _should_ work because `platformStatus` _should_ always be set on a GCP cluster (`platformStatus` wasn't set on 4.1 clusters, but 4.1 couldn't be installed on GCP, so it should be impossible to have a GCP cluster that was upgraded from 4.1).  However, it is a pointer type, and the value is technically optional, so I'd still recommend having a nil check to be safe.

alebedev87 replied:
Done

Miciah commented on api/v1alpha1/externaldns_utils.go:
```diff
@@ -0,0 +1,26 @@
+
+package v1alpha1
+
+// PublicProvider returns true if the dns provider is one of supported public providers
+func (e *ExternalDNS) PublicProvider() bool {
```
Is "Public" significant here, or would it be reasonable to name this method something like `ValidProvider` or `SupportedProvider`?

alebedev87 replied:
No, not significant, I just didn't come up with a better name. The idea of this method is to tell that "if you are in the OpenShift case you need to do something only if it's one of these providers". Because we don't leverage OpenShift features for `BlueCat` and `InfoBlox`.

`Supported`, I don't know, this may give an impression that `BlueCat` and `InfoBlox` are not supported. The same can be thought about `Valid`. WDYT? Do you have any other ideas?

arjunrn replied:
Maybe `AutoCredentialConfigProvider()` if the cloud provider is supported for automatic credential configuration.

alebedev87 replied:
@arjunrn: sorry, didn't see your comment. I called it `ManagedCredentialsProvider()`. Also I moved it outside the api package to remove the unnecessary method from the "public" API.

Miciah reviewed (commented):
Very nice cleanup!  A couple of minor suggestions, but overall
/approve

alebedev87 commented:
/assign @miheer

alebedev87 commented:
aws e2e test failed: multiple nameservers failed to find the record. Intermittent?

/retest

alebedev87 commented:
AWS e2e failure is due to the operand POD which didn't find a key in the credentials secret:
```yaml
                "containerStatuses": [
                    {
                        "image": "k8s.gcr.io/external-dns/external-dns:v0.8.0",
                        "imageID": "",
                        "lastState": {},
                        "name": "external-dns-n587hc8h58fh55cq",
                        "ready": false,
                        "restartCount": 0,
                        "started": false,
                        "state": {
                            "waiting": {
                                "message": "couldn't find key aws_access_key_id in Secret external-dns/external-dns-credentials-test-extdns",
                                "reason": "CreateContainerConfigError"
                            }
                        }
                    }
                ],
```

miheer commented:
@alebedev87 please make sure you test this on Azure, AWS and GCP.

miheer commented:
> AWS e2e failure is due to the operand POD which didn't find a key in the credentials secret:
> 
> ```yaml
>                 "containerStatuses": [
>                     {
>                         "image": "k8s.gcr.io/external-dns/external-dns:v0.8.0",
>                         "imageID": "",
>                         "lastState": {},
>                         "name": "external-dns-n587hc8h58fh55cq",
>                         "ready": false,
>                         "restartCount": 0,
>                         "started": false,
>                         "state": {
>                             "waiting": {
>                                 "message": "couldn't find key aws_access_key_id in Secret external-dns/external-dns-credentials-test-extdns",
>                                 "reason": "CreateContainerConfigError"
>                             }
>                         }
>                     }
>                 ],
> ```

Yes please test the code manually on AWS.

miheer commented:
You need to make changes here

https://github.com/openshift/external-dns-operator/blob/main/pkg/operator/controller/externaldns/controller.go#L108

miheer commented:
Also here

https://github.com/openshift/external-dns-operator/blob/main/pkg/operator/controller/credentials-secret/controller.go#L256

miheer commented:
I think you have made the above changes but it looks somewhere the bool is not getting set properly which is not creating the secret properly.

miheer commented:
I will really recommend to test this manually on AWS, Azure and GCP.

arjunrn commented on api/v1alpha1/externaldns_webhook.go:
```diff
@@ -17,60 +17,26 @@ limitations under the License.
-	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
-		Complete()
+func (r *ExternalDNS) SetupWebhookWithManager(mgr ctrl.Manager, openshift bool) error {
```
It may be time to end this hacky way in which kubebuilder insists on registering the webhook. In a PR once this is merged I think I would like to implement a struct which embeds `ExternalDNS` and has the `isOpenshift` field. This way we can avoid the package level variable here as well.

alebedev87 commented:
@miheer : there is no need to test the code manually on any provider. CCO logic is already tested by the e2e tests. That's actually why I got the error which I fixed in the commit that followed the comment: because I modified `desiredCredentialsSecret` function which was used in the CCO related flow.

alebedev87 commented:
Added the change to not use the CCO credentials if they are explicitly specified in CR.

Use case when credentials are taken from CCO will be covered in https://github.com/openshift/external-dns-operator/pull/82.

Miciah commented on pkg/operator/config/config.go:
```diff
@@ -55,16 +72,70 @@ type Config struct {
+	// IsOpenShift is the flag indicating that the operator runs in OpenShift cluster
+	IsOpenShift bool
+
+	// PlatformStatus is the details about the undelying platform
```
```suggestion
	// PlatformStatus is the details about the underlying platform.
```

Miciah commented on pkg/operator/config/config.go:
```diff
@@ -17,17 +17,34 @@ limitations under the License.
+	DefaultOperandNamespace        = "external-dns"
+	DefaultEnableWebhook           = true
+	DefaultEnablePlatformDetection = true
+	DefaultIsOpenShift             = false
```
`DefaultIsOpenShift` can be removed too along with the `operatorconfig.New` function.

Miciah commented on pkg/utils/utils.go:
```diff
@@ -0,0 +1,26 @@
+/*
+You may obtain a copy of the License at
```
Looks like the licence text was truncated.  
```suggestion
/*
Copyright 2021.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
```

Miciah commented:
/lgtm
